### ISSUE TITLE
keras explanations with dataframe data

### ISSUE THREAD

**Issue by dhicks** [thomasp85/lime#139]
I'm having some issues using `lime` with `keras` when the data takes the form of a matrix/dataframe. Briefly, there's no `lime()` method for matrices, so matrix data needs to be passed in as a dataframe.  But the `predict_model()` method for Keras models calls `as.array(newdata)`, which raises the error
```
Error in `dimnames<-.data.frame`(`*tmp*`, value = list(n)) : 
  invalid 'dimnames' given for data frame
```

Here's a MWE for a Keras model with this problem, based on the [Getting Started vignette](https://keras.rstudio.com/articles/getting_started.html).  (I'm just getting started with Keras, so I apologize that this isn't as minimal as it could be.)  

```
library(keras)

mnist <- dataset_mnist()
x_train <- mnist$train$x
y_train <- mnist$train$y
x_test <- mnist$test$x
y_test <- mnist$test$y

# reshape
x_train <- array_reshape(x_train, c(nrow(x_train), 784))
x_test <- array_reshape(x_test, c(nrow(x_test), 784))
# rescale
x_train <- x_train / 255
x_test <- x_test / 255

y_train <- to_categorical(y_train, 10)
y_test <- to_categorical(y_test, 10)

model <- keras_model_sequential() 
model %>% 
    layer_dense(units = 256, activation = 'relu', input_shape = c(784)) %>% 
    layer_dropout(rate = 0.4) %>% 
    layer_dense(units = 128, activation = 'relu') %>%
    layer_dropout(rate = 0.3) %>%
    layer_dense(units = 10, activation = 'softmax')

model %>% 
    compile(
        loss = 'categorical_crossentropy',
        optimizer = optimizer_rmsprop(),
        metrics = c('accuracy')
    )

history <- model %>% 
    fit(
        x_train, y_train, 
        epochs = 10, batch_size = 128, 
        validation_split = 0.2
    )
```

Given the environment created by running that code, the following should reproduce the error:  
```
library(lime)

explainer = lime(as.data.frame(x_train), as_classifier(model))
lime::explain(head(as.data.frame(x_train)), 
              explainer,
              n_labels = 1, n_features = 1)
```

**Comment by aliawadh980** [comment]
I had the same error, and I managed to solve it by overwriting the correct output of class(model). It seems the newer version uses different name. So, it is fixed for by:
```{r}
class(model_keras)
# [1] "keras.engine.sequential.Sequential"
# [2] "keras.engine.training.Model"       
# [3] "keras.engine.network.Network"      
# [4] "keras.engine.base_layer.Layer"     
# [5] "python.builtin.object"    
```
Then use `keras.engine.sequential.Sequential` instead of `keras.models.Sequential`, i.e.:
```{r}
# gives error
model_type.keras.models.Sequential <- function(x, ...) {
  "classification"
}

# works!
model_type.keras.engine.sequential.Sequential <- function(x, ...) {
  "classification"
}

```

The same for `predict_model()`, use `keras.engine.sequential.Sequential`, or what is the output of `class(model)`

**Comment by dhicks** [comment]
@aliawadh980 I don't understand the workaround you're suggesting.  Assigning the second function, both before and after loading lime, produces the same error for me.

**Comment by aliawadh980** [comment]
I was having the same error as yours. So what I have done is:
1. Check the output of `class(model_keras)`
2. Use the correct output name from step 1 to overwrite the two functions: `model_type` and `predict_model`
3. Then test the code `predict_model(x = rna, newdata = (train_set), type = 'raw')`
This worked for me, and I hope it works for you too!

Please note: I have the latest libraries installed.

**Comment by dhicks** [comment]
@aliawadh980 could you create a self-contained example?  I'm using `keras` 2.2.4 and `lime` 0.4.1, so that's not the issue.  @felipefernandesmoreira 's example uses a dataset `train_set` that isn't included here and doesn't seem to be in the link he gives, so I can't run it to check your proposed solution.  My `class(model)` has exactly the same results as your `class(model_keras)`; but every way I tried to implement your proposed solution gave an error.  So it seems like either I don't understand how to implement your proposal or your proposal doesn't work in my environment for some reason.  A self-contained example would be the easiest way to check the first possibility.

**Comment by felipefernandesmoreira** [comment]
Holy cow, that worked @aliawadh980

**Comment by grcasey** [comment]
> I had the same error, and I managed to solve it by overwriting the correct output of class(model). It seems the newer version uses different name. So, it is fixed for by:
> 
> ```
> class(model_keras)
> # [1] "keras.engine.sequential.Sequential"
> # [2] "keras.engine.training.Model"       
> # [3] "keras.engine.network.Network"      
> # [4] "keras.engine.base_layer.Layer"     
> # [5] "python.builtin.object"    
> ```
> Then use `keras.engine.sequential.Sequential` instead of `keras.models.Sequential`, i.e.:
> 
> ```
> # gives error
> model_type.keras.models.Sequential <- function(x, ...) {
>   "classification"
> }
> 
> # works!
> model_type.keras.engine.sequential.Sequential <- function(x, ...) {
>   "classification"
> }
> ```
> The same for `predict_model()`, use `keras.engine.sequential.Sequential`, or what is the output of `class(model)`

Thanks so much!

**Comment by dhicks** [comment]
@thomasp85 Is this resolved in a pending release?  As I noted above, every way I tried to implement @aliawadh980 's proposed solution gave me an error, and they never supplied a self-contained example to explain the solution exactly.  So this isn't hasn't been resolved with `lime` 0.4.1.

**Comment by thomasp85** [comment]
I mistook the issue as solved... will look into it and provide a fix

**Comment by felipefernandesmoreira** [comment]
I merged the modeling code posted by @dhicks with my explainer code and ran without the above error - now it says that "it cannot allocate vector of size 381.5 Mb". Don't know what that means, but I guess it is an advancement. 

```
library(lime)

# Setup lime::model_type()
model_type.keras.engine.sequential.Sequential <- function(x, ...) {
  "classification"}

# Setup lime::predict_model()
predict_model.keras.engine.sequential.Sequential <- function (x, newdata, type, ...) {
  pred <- predict(object = x, x = as.matrix(newdata))
  data.frame (Positive = pred, Negative = 1 - pred) }

# Teste do modelo
predict_model (x       = model, 
               newdata = (x_train), 
               type    = 'raw')


# Usando lime() no train set
explainer <- lime(x = as.data.frame(x_train), model= model)

# Rodando o Explainer

system.time (
  explanation <- lime::explain (
    x = as.data.frame(x_test),
    explainer    = explainer, 
    n_features = 1,
    n_labels = 1))``
```

**Comment by felipefernandesmoreira** [comment]
Did some changes for the purpose of testing and it freaking ran.

```
library(lime)

# Setup lime::model_type()
model_type.keras.engine.sequential.Sequential <- function(x, ...) {
  "classification"}

# Setup lime::predict_model()
predict_model.keras.engine.sequential.Sequential <- function (x, newdata, type, ...) {
  pred <- predict(object = x, x = as.matrix(newdata))
  data.frame (Positive = pred, Negative = 1 - pred) }

# Teste do modelo
predict_model (x       = model, 
               newdata = (x_train), 
               type    = 'raw')


# Usando lime() no train set
explainer <- lime(x = as.data.frame(x_train), model= model)

# Rodando o Explainer

system.time (
  explanation <- lime::explain (
    x = as.data.frame(x_test),
    explainer    = explainer, 
    n_features = 1,
    n_labels = 1))
```
The result of `plot_explanations(explanation)` is below (no idea of what that means, but hey, there it is).
![explain_IMG](https://user-images.githubusercontent.com/5262199/59375244-16100b00-8d24-11e9-991f-9a762faae3b6.png)

**Comment by thomasp85** [comment]
Thanks for taking the time, @felipefernandesmoreira — @dhicks can you confirm that the code works for you?

**Comment by dhicks** [comment]
I'm running up against memory limits on my little old laptop, and don't have the time in the immediate future to put together a smaller test case or spin up a more powerful machine on AWS.  But I'm also not getting the error from calling `as.array()`, and running up against the memory limits indicates that this is actually getting to the LIME logic.  So I'm okay with closing this issue until it appears again, if ever.

**Comment by CarstenLange** [comment]
Thank you so much. Wasted half a day before..

**Comment by CarstenLange** [comment]
Somebody asked for a working model. here is one for regression:
library(lime)
library(keras)

AllData=tibble(x1=c(1:1000),x2=c(1:1000),y=x1^0.4*x2^0.6+
                 runif(1,-100,100))
TestIndex=sample(1:1000,100)
xTrain=scale(data.matrix(AllData[-TestIndex,1:2]))
yTrain=data.matrix(AllData[-TestIndex,3])
xTest=scale(data.matrix(AllData[TestIndex,1:2]))
yTest=data.matrix(AllData[TestIndex,3])
## @knitr RunNn
use_session_with_seed(2019)
FLAGS<-flags(flag_integer("dense1Neurons",10),
             flag_integer("epochs",100)
             ,flag_numeric("dense1Dropout",0.5))

ModelNn <- keras_model_sequential()
ModelNn %>%
  layer_dense(units = FLAGS$dense1Neurons , activation = 'relu', input_shape = c(2)) %>%
  layer_dropout(rate = FLAGS$dense1Dropout)%>%
  layer_dense(units = 1, activation = 'linear')

ModelNn %>% compile(loss='mean_squared_error', optimizer='Adam', 
                    metrics='mape')
history <- ModelNn %>% fit(
  xTrain, yTrain,validation_data=list(xTest,yTest),
  epochs=FLAGS$epochs, batch_size=5, verbose=1)
plot(history)
summary(ModelNn)


model_type.keras.engine.sequential.Sequential <- function(x, ...) {
  "regression"
  }
predict_model.keras.engine.sequential.Sequential <- 
  function (x, newdata, type, ...) {
  pred=predict(x, as.matrix(newdata))
  as.data.frame(pred)
   }


ExplainerModelNn=lime(as.tibble(xTrain),ModelNn)
ExplanationModelNn <- explain(as.data.frame(xTest[1:20,]), 
                              ExplainerModelNn, 
                                      n_features = 2)
plot_features(ExplanationModelNn)
plot_explanations(ExplanationModelNn)

**Comment by neural-oracle** [comment]
save my day
